# Evaluating The Performance of Various Bayesian Models

In [1]:
import fourtenmodule
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.feature_selection import VarianceThreshold


## Synthetic Data

### Initial Set-Up

In [2]:
np.random.seed(410)
n=50
p = 5
# Create alternating columns of [0, 1] and [1, 0] in 3 columns to reduce experimental randomness
X = np.ones((n, p))
for i in range(0, X.shape[1], 2):
    X[i::2, i] = 0
    X[i+1::2, i] = 1

# Create true_beta based on logspace
true_beta = np.logspace(0, 1, p, base=2)
sigma_true = 1
print(true_beta)

# Generate y with added random noise
y = np.dot(X, true_beta) + stats.norm(0,sigma_true).rvs(n)

fourtenmodule.run_and_plot_models(X, y, true_beta, n_components=2)

Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.


[1.         1.18920712 1.41421356 1.68179283 2.        ]

=== Running timing tests (max 5s per model) ===

Testing Bayesian Linear Regression...


Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 2.98s

Testing Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 1.28s

Testing Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.58s

Testing Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.


  → Passed in 0.99s

Testing Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 83.079
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 1.44s

Testing Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/decomposition/_fastica.py:629: RuntimeWarning: invalid value encountered in divide
  K = (u / d).T[:n_components]  # see (6.33) p.140
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.51s

Testing Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.65s

=== Running full analysis on passing models ===

Running Bayesian Linear Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 7 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Linear Regression successfully

Running Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 12 seconds.
There were 215 divergences after tuning. Increase `target_accept` or reparameterize.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Ridge Regression successfully

Running Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 47 seconds.
There were 103 divergences after tuning. Increase `target_accept` or reparameterize.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Lasso successfully

Running Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 10 seconds.


Completed Bayesian Robust Regression successfully

Running Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 83.523
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Variational Inference successfully

Running Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/decomposition/_fastica.py:629: RuntimeWarning: invalid value encountered in divide
  K = (u / d).T[:n_components]  # see (6.33) p.140
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian PCR successfully

Running Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.


Completed Bayesian ICR successfully

=== Final Results (only passing models) ===

Model                          RMSE       Beta Estimates
----------------------------------------------------------------------
Bayesian Lasso                 0.8930     β0:0.796  β1:1.328  β2:0.811  β3:1.653  β4:2.799
Bayesian Linear Regression     0.8954     β0:0.808  β1:1.249  β2:1.225  β3:1.743  β4:2.365
Bayesian Robust Regression     0.8954     β0:0.802  β1:1.256  β2:1.227  β3:1.743  β4:2.362
Bayesian Variational Inference 0.9001     β0:0.957  β1:1.261  β2:1.398  β3:1.732  β4:2.047
Bayesian Ridge Regression      0.9018     β0:0.978  β1:1.224  β2:1.403  β3:1.713  β4:2.049
Bayesian ICR                   4.0488     β0:0.266  β1:0.000  β2:0.921  β3:0.000  β4:1.549
Bayesian PCR                   4.2863     β0:0.765  β1:0.000  β2:0.803  β3:0.000  β4:0.819


{'passing_models': {'Bayesian Linear Regression': <function fourtenmodule.bayesian_regression_mcmc(X, y, true_beta=None, draws=4000)>,
  'Bayesian Ridge Regression': <function fourtenmodule.bayesian_ridge_regression(X, y, true_beta=None, draws=4000)>,
  'Bayesian Lasso': <function fourtenmodule.bayesian_lasso(X, y, true_beta=None, n_folds=5, draws=4000)>,
  'Bayesian Robust Regression': <function fourtenmodule.bayesian_robust_regression(X, y, true_beta=None, draws=4000)>,
  'Bayesian Variational Inference': <function fourtenmodule.bayesian_regression_vi(X, y, true_beta=None)>,
  'Bayesian PCR': <function fourtenmodule.bayesian_pcr(X, y, true_beta=None, n_components=None, draws=4000)>,
  'Bayesian ICR': <function fourtenmodule.bayesian_icr(X, y, true_beta=None, n_components=None, draws=4000)>},
 'results': [{'Model': 'Bayesian Linear Regression',
   'RMSE': np.float64(0.8953929670557731),
   'Beta Estimates': array([0.80798536, 1.24876424, 1.22502916, 1.74342438, 2.36533822])},
  {'Mode

The true beta is given by [1, 1.18920712, 1.41421356, 1.68179283, 2]. The best model by RMSE appears to be the Bayesian Lasso Regression, with Linear and Robust regression producing similar scores. On the other hand, ICR and PCR produce significantly worse RMSE values due to the small number of predictors.


## Changing Parameters

### Sample Size

In [ ]:
# Define sample sizes to test
sample_sizes = [50, 100, 500, 1000, 1500]
rmse_results = {"MCMC": [], "Ridge": [], "Lasso": [], "Robust": [], "VI": [], "PCR" : [], "ICR": []}

for n in sample_sizes:
    print(f"Running models for sample size: {n}")
    rmse_mcmc, rmse_ridge, rmse_lasso, rmse_robust, rmse_vi, rmse_pcr, rmse_icr = fourtenmodule.run_models_and_evaluate(n = n)
    rmse_results["MCMC"].append(rmse_mcmc)
    rmse_results["Ridge"].append(rmse_ridge)
    rmse_results["Lasso"].append(rmse_lasso)
    rmse_results["Robust"].append(rmse_robust)
    rmse_results["VI"].append(rmse_vi)
    rmse_results["PCR"].append(rmse_pcr)
    rmse_results["ICR"].append(rmse_icr)

# Plot RMSE as a function of sample size
plt.figure(figsize=(8, 6))
for model, rmse in rmse_results.items():
    plt.plot(sample_sizes, rmse, marker='o', label=model)
plt.xlabel("Sample Size")
plt.ylabel("RMSE of Estimated Beta")
plt.title("Model Performance as Sample Size Increases")
plt.legend()
plt.grid()
plt.show()

In [ ]:
plt.figure(figsize=(8, 6))

for model, rmse in rmse_results.items():
    if model != "ICR":
        plt.plot(sample_sizes, rmse, marker='o', label=model)
        
plt.xlabel("Sample Size")
plt.ylabel("RMSE of Estimated Beta")
plt.title("Model Performance as Sample Size Increases")
plt.legend()
plt.grid()
plt.show()

### Number of parameters

In [ ]:
# Define sample sizes to test
p_numbers = [5, 20, 30, 60, 100]
n=20
rmse_results = {"MCMC": [], "Ridge": [], "Lasso": [], "Robust": [], "VI": [], "PCR" : [], "ICR": []}

for p in p_numbers:
    print(f"Running models for this many parameters: {p}")
    rmse_mcmc, rmse_ridge, rmse_lasso, rmse_robust, rmse_vi, rmse_pcr, rmse_icr = fourtenmodule.run_models_and_evaluate(p = p)
    rmse_results["MCMC"].append(rmse_mcmc)
    rmse_results["Ridge"].append(rmse_ridge)
    rmse_results["Lasso"].append(rmse_lasso)
    rmse_results["Robust"].append(rmse_robust)
    rmse_results["VI"].append(rmse_vi)
    rmse_results["PCR"].append(rmse_pcr)
    rmse_results["ICR"].append(rmse_icr)

# Plot RMSE as a function of sample size
plt.figure(figsize=(8, 6))
for model, rmse in rmse_results.items():
    plt.plot(p_numbers, rmse, marker='o', label=model)
plt.xlabel("Number of Predictors")
plt.ylabel("RMSE of Estimated Beta")
plt.title("Model Performance as Number of Predictors Increases")
plt.legend()
plt.grid()
plt.show()

### Using highly sparse data

In [ ]:
# Define sample sizes to test
p_numbers = [5, 20, 30, 60, 100]
n=20
rmse_results = {"MCMC": [], "Ridge": [], "Lasso": [], "Robust": [], "VI": [], "PCR" : [], "ICR": []}

for p in p_numbers:
    print(f"Running models for this many parameters: {p}")
    rmse_mcmc, rmse_ridge, rmse_lasso, rmse_robust, rmse_vi, rmse_pcr, rmse_icr = fourtenmodule.run_models_and_evaluate(p = p, high_corr = True)
    rmse_results["MCMC"].append(rmse_mcmc)
    rmse_results["Ridge"].append(rmse_ridge)
    rmse_results["Lasso"].append(rmse_lasso)
    rmse_results["Robust"].append(rmse_robust)
    rmse_results["VI"].append(rmse_vi)
    rmse_results["PCR"].append(rmse_pcr)
    rmse_results["ICR"].append(rmse_icr)

# Plot RMSE as a function of sample size
plt.figure(figsize=(8, 6))
for model, rmse in rmse_results.items():
    plt.plot(p_numbers, rmse, marker='o', label=model)
plt.xlabel("Number of Predictors")
plt.ylabel("RMSE of Estimated Beta")
plt.title("Model Performance as Number of Predictors Increases With Sparse True Signals")
plt.legend()
plt.grid()
plt.show()

### Magnitudes

#### Beta 0

In [ ]:
# Define different values where only one coefficient increases at a time
p=5
# Fixed 5-element true_beta array
true_beta = np.logspace(0, 1, p, base=2)

beta_variations = [[2**i, *true_beta[1:]] for i in range(5)]


n = 20
rmse_results = {"MCMC": [], "Ridge": [], "Lasso": [], "Robust": [], "VI": [], "PCR": [], "ICR": []}

for true_beta in beta_variations:
    print(f"Running models with true_beta: {true_beta}")
    rmse_mcmc, rmse_ridge, rmse_lasso, rmse_robust, rmse_vi, rmse_pcr, rmse_icr = fourtenmodule.run_models_and_evaluate(n = n, true_beta = true_beta)
    rmse_results["MCMC"].append(rmse_mcmc)
    rmse_results["Ridge"].append(rmse_ridge)
    rmse_results["Lasso"].append(rmse_lasso)
    rmse_results["Robust"].append(rmse_robust)
    rmse_results["VI"].append(rmse_vi)
    rmse_results["PCR"].append(rmse_pcr)
    rmse_results["ICR"].append(rmse_icr)

# Plot RMSE as a function of increasing individual beta values
plt.figure(figsize=(8, 6))
beta_labels = [
    "1", "2", "4", "8", "16"
]
for model, rmse in rmse_results.items():
    plt.plot(beta_labels, rmse, marker='o', label=model)
plt.xlabel("True Beta 0 Variations")
plt.ylabel("RMSE of Estimated Beta 0")
plt.title("Model Performance as Beta 0 Increases")
plt.xticks(rotation=45)
plt.legend()
plt.grid()
plt.show()


#### Beta 1

In [ ]:
p=5
# Fixed 5-element true_beta array
true_beta = np.logspace(0, 1, p, base=2)

# Define different values where only one coefficient increases at a time
beta_variations = [ [true_beta[0], 2**i, *true_beta[2:]] for i in range(5)]


n = 20
rmse_results = {"MCMC": [], "Ridge": [], "Lasso": [], "Robust": [], "VI": [], "PCR": [], "ICR": []}

for true_beta in beta_variations:
    print(f"Running models with true_beta: {true_beta}")
    rmse_mcmc, rmse_ridge, rmse_lasso, rmse_robust, rmse_vi, rmse_pcr, rmse_icr = fourtenmodule.run_models_and_evaluate(n = n, true_beta = true_beta)
    rmse_results["MCMC"].append(rmse_mcmc)
    rmse_results["Ridge"].append(rmse_ridge)
    rmse_results["Lasso"].append(rmse_lasso)
    rmse_results["Robust"].append(rmse_robust)
    rmse_results["VI"].append(rmse_vi)
    rmse_results["PCR"].append(rmse_pcr)
    rmse_results["ICR"].append(rmse_icr)

# Plot RMSE as a function of increasing individual beta values
plt.figure(figsize=(8, 6))
beta_labels = [
    "1", "1.41", "4", "8", "16"
]
for model, rmse in rmse_results.items():
    plt.plot(beta_labels, rmse, marker='o', label=model)
plt.xlabel("True Beta 1 Variations")
plt.ylabel("RMSE of Estimated Beta 1")
plt.title("Model Performance as Beta 1 Increases")
plt.xticks(rotation=45)
plt.legend()
plt.grid()
plt.show()


# Real Data

## Housing Data

In [4]:
def load_soccer_data():
    try:
        url = "https://raw.githubusercontent.com/javmencia/STA410/refs/heads/main/housing.csv"
        return pd.read_csv(url, header=None, sep=r"\s+")
    except:
        try:
            return pd.read_csv("housing.csv", header=None, sep=r"\s+")
        except:
            raise Exception("Could not load data")

housing = load_soccer_data()
# Load dataset
housing.columns = ["CRIM", "ZN", "INDUS", "CHAS", "NOX", "RM", "AGE", "DIS", "RAD", "TAX", "PTRATIO", "B", "LSTAT", "MEDV"]

# Define X (predictors) and y (target variable)
X = housing.drop(columns=["MEDV"]).values
y = housing["MEDV"].values

# Normalize X
X = (X - X.mean(axis=0)) / X.std(axis=0)



results = fourtenmodule.run_and_plot_models(X, y, n_components=2)




Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...



=== Running timing tests (max 5s per model) ===

Testing Bayesian Linear Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.98s

Testing Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 1 seconds.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 1.83s

Testing Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.92s

Testing Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 1 seconds.
The number of samples is too small to check convergence reliably.


  → Passed in 1.23s

Testing Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 2,405
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 1.52s

Testing Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.75s

Testing Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.71s

=== Running full analysis on passing models ===

Running Bayesian Linear Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 6 seconds.


Completed Bayesian Linear Regression successfully

Running Bayesian Ridge Regression...


Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 14 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Ridge Regression successfully

Running Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 6 seconds.


Completed Bayesian Lasso successfully

Running Bayesian Robust Regression...


Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 6 seconds.


Completed Bayesian Robust Regression successfully

Running Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 2,424.8
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Variational Inference successfully

Running Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian PCR successfully

Running Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.


Completed Bayesian ICR successfully

=== Final Results (only passing models) ===

Model                          RMSE       Beta Estimates
----------------------------------------------------------------------
Bayesian Lasso                 23.0218    β0:-0.724  β1:0.780  β2:-0.213  β3:0.690  β4:-1.462  β5:2.776  β6:-0.070  β7:-2.460  β8:1.479  β9:-1.100  β10:-1.843  β11:0.780  β12:-3.562
Bayesian Ridge Regression      23.0997    β0:-0.599  β1:0.489  β2:-0.441  β3:0.682  β4:-0.669  β5:2.485  β6:-0.214  β7:-1.130  β8:0.340  β9:-0.517  β10:-1.450  β11:0.692  β12:-2.492
Bayesian Variational Inference 23.1627    β0:-0.632  β1:0.559  β2:-0.599  β3:0.804  β4:-0.530  β5:2.061  β6:-0.294  β7:-0.837  β8:0.117  β9:-0.496  β10:-1.432  β11:0.714  β12:-1.949
Bayesian Linear Regression     23.2191    β0:-0.509  β1:0.412  β2:-0.486  β3:0.579  β4:-0.446  β5:1.965  β6:-0.263  β7:-0.490  β8:-0.058  β9:-0.462  β10:-1.125  β11:0.551  β12:-1.809
Bayesian Robust Regression     23.2272    β0:-0.501  β1:0.405

## Soccer Players' Market Value
Many redundant and multicollinear variables, also too many features, expect most algorithms to time out and ICR or PCR to perform better

In [5]:
# Load dataset (with SSL fix)
def load_soccer_data():
    try:
        url = "https://raw.githubusercontent.com/javmencia/STA410/main/transfermarkt201718.csv"
        return pd.read_csv(url, sep=";")
    except:
        try:
            return pd.read_csv("transfermarkt201718.csv", sep=";")
        except:
            raise Exception("Could not load data")

soccer = load_soccer_data()


print(len(soccer))
# Preprocess data and select top 100 features
X, y = fourtenmodule.preprocess_data(soccer, "value")


# Now run models
results = fourtenmodule.run_and_plot_models(X, y)



Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.


2232

Selected top 2232 features from 2837 total features
Top 10 features by variance:
- player_Jonjoe Kenny
- player_Kevin Akpoguma
- player_Gerson
- player_Cuco Martina
- nationality_cw CUW
- player_Benjamin Hübner
- player_Alfred Duncan
- player_Kevin Mirallas
- player_Oumar Niasse
- player_Jordan Pickford

=== Running timing tests (max 5s per model) ===

Testing Bayesian Linear Regression...


Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 4 seconds.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.


  → Failed: took 5.54s (timeout)

Testing Bayesian Ridge Regression...


Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 534 seconds.
The number of samples is too small to check convergence reliably.


  → Failed: took 537.84s (timeout)

Testing Bayesian Lasso...


/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 5.037e+13, tolerance: 4.033e+13
  model = cd_fast.enet_coordinate_descent(
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.128e+14, tolerance: 4.033e+13
  model = cd_fast.enet_coordinate_descent(
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iter

Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 1 seconds.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Failed: took 187.73s (timeout)

Testing Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 209 seconds.
The number of samples is too small to check convergence reliably.


  → Failed: took 210.39s (timeout)

Testing Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 1.1311e+15
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Failed: took 109.93s (timeout)

Testing Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.


  → Passed in 0.55s

Testing Bayesian ICR...


Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.


  → Passed in 2.98s

=== Models that didn't pass timing tests ===
Bayesian Linear Regression: Too slow (5.54s > 5s)
Bayesian Ridge Regression: Too slow (537.84s > 5s)
Bayesian Lasso: Too slow (187.73s > 5s)
Bayesian Robust Regression: Too slow (210.39s > 5s)
Bayesian Variational Inference: Too slow (109.93s > 5s)

=== Running full analysis on passing models ===

Running Bayesian PCR...


Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 4 seconds.


Completed Bayesian PCR successfully

Running Bayesian ICR...


Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 4 seconds.


Completed Bayesian ICR successfully

=== Final Results (only passing models) ===

Model                          RMSE       Beta Estimates
----------------------------------------------------------------------
Bayesian ICR                   19246890.6483 β0:-0.000  β1:-0.000  β2:-0.000  β3:-0.000  β4:-0.000  β5:-0.000  β6:-0.000  β7:0.000  β8:-0.000  β9:0.000  β10:-0.000  β11:-0.000  β12:-0.000  β13:0.000  β14:-0.000  β15:-0.000  β16:-0.000  β17:-0.000  β18:-0.000  β19:-0.000  β20:-0.000  β21:-0.000  β22:0.000  β23:-0.000  β24:0.000  β25:-0.000  β26:-0.000  β27:-0.000  β28:0.000  β29:-0.000  β30:0.000  β31:0.000  β32:0.000  β33:-0.000  β34:0.000  β35:-0.000  β36:0.000  β37:0.000  β38:-0.000  β39:-0.000  β40:-0.000  β41:-0.000  β42:-0.000  β43:-0.000  β44:0.000  β45:0.000  β46:-0.000  β47:-0.000  β48:-0.000  β49:-0.000  β50:-0.000  β51:-0.000  β52:0.000  β53:0.000  β54:-0.000  β55:-0.000  β56:-0.000  β57:-0.000  β58:0.000  β59:0.000  β60:0.000  β61:-0.000  β62:-0.000  β63:0.000  β64:-0.

### Cancer Survival Data
Highly collinear

In [2]:
def load_cancer_data():
    try:
        url = "https://raw.githubusercontent.com/javmencia/STA410/refs/heads/main/METABRIC_RNA_Mutation.csv"
        return pd.read_csv(url, sep=";")
    except:
        try:
            return pd.read_csv("METABRIC_RNA_Mutation.csv")
        except:
            raise Exception("Could not load data")

cancer = load_cancer_data()
# Load dataset

# Define X (predictors) and y (target variable)

X, y = fourtenmodule.preprocess_data(cancer, "overall_survival")

# Normalize X
X = (X - X.mean(axis=0)) / X.std(axis=0)



results = fourtenmodule.run_and_plot_models(X, y, n_components=2)




/var/folders/x9/3zl22m596tl6d44cbs62z6h00000gn/T/ipykernel_23385/1295189474.py:7: DtypeWarning: Columns (678,688,690,692) have mixed types. Specify dtype option on import or set low_memory=False.
  return pd.read_csv("METABRIC_RNA_Mutation.csv")
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.



Selected top 692 features from 8658 total features
Top 10 features by variance:
- pbrm1_mut_0
- er_status_measured_by_ihc_nan
- l1cam_mut_0
- tumor_size_14.0
- ncoa3_mut_0
- bap1_mut_0
- kdm3a_mut_0
- sbno1_mut_0
- lifr_mut_0
- aff2_mut_0

=== Running timing tests (max 5s per model) ===

Testing Bayesian Linear Regression...


Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 8 seconds.
There were 43 divergences after tuning. Increase `target_accept` or reparameterize.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Failed: took 12.39s (timeout)

Testing Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 31 seconds.
There were 59 divergences after tuning. Increase `target_accept` or reparameterize.
The number of samples is too small to check convergence reliably.


  → Failed: took 33.14s (timeout)

Testing Bayesian Lasso...


Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
There were 55 divergences after tuning. Increase `target_accept` or reparameterize.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 1.95s

Testing Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 9 seconds.
There were 37 divergences after tuning. Increase `target_accept` or reparameterize.
Chain 1 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
The number of samples is too small to check convergence reliably.


  → Failed: took 11.22s (timeout)

Testing Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 7,626
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Failed: took 7.58s (timeout)

Testing Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
There were 2 divergences after tuning. Increase `target_accept` or reparameterize.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.56s

Testing Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.


  → Passed in 0.59s

=== Models that didn't pass timing tests ===
Bayesian Linear Regression: Too slow (12.39s > 5s)
Bayesian Ridge Regression: Too slow (33.14s > 5s)
Bayesian Robust Regression: Too slow (11.22s > 5s)
Bayesian Variational Inference: Too slow (7.58s > 5s)

=== Running full analysis on passing models ===

Running Bayesian Lasso...


Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 734 draw iterations (4_000 + 2_936 draws total) took 665 seconds.
Chain 0 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 1 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 2 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 3 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Lasso successfully

Running Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian PCR successfully

Running Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.


Completed Bayesian ICR successfully

=== Final Results (only passing models) ===

Model                          RMSE       Beta Estimates
----------------------------------------------------------------------
Bayesian Lasso                 0.4272     β0:0.002  β1:-1.750  β2:0.001  β3:0.002  β4:-0.001  β5:-0.001  β6:0.006  β7:0.001  β8:-0.006  β9:0.001  β10:-0.000  β11:-0.000  β12:-0.007  β13:-0.006  β14:0.003  β15:-0.002  β16:-0.004  β17:-0.007  β18:-0.004  β19:-0.001  β20:-0.004  β21:0.006  β22:0.003  β23:0.001  β24:-0.002  β25:0.000  β26:0.000  β27:0.002  β28:-0.004  β29:-0.003  β30:0.002  β31:0.001  β32:0.008  β33:-0.001  β34:-0.006  β35:0.003  β36:0.007  β37:-0.001  β38:0.001  β39:0.001  β40:-0.001  β41:-0.007  β42:-0.003  β43:0.006  β44:-0.001  β45:-0.003  β46:-0.003  β47:-0.001  β48:0.000  β49:-0.002  β50:-0.007  β51:-0.002  β52:0.001  β53:-0.000  β54:-0.384  β55:0.005  β56:-0.011  β57:0.002  β58:0.002  β59:0.004  β60:0.001  β61:-0.005  β62:-0.007  β63:0.004  β64:0.004  β65:-0.0

In [ ]:
print(results['all_results'])

